openslp.misc/debian/slptool.1
.TH slptool 1 "Sep 08 2001" "OpenSLP 1.0.1"

.SH NAME
slptool \- SLP command line tool

.SH SYNOPSIS
.B /usr/bin/slpd 
.B [\-v|\-\-version] 
.B [\-s|\-\-scope scopes]
.B [\-l|\-\-language languagetag] 
.B [command [arguments]]
.B 

.SH DESCRIPTION
.LP
.B slptool
is a command line tool to make SLPv2 User Agent (UA) requests. It can be
used for SLP lookup/advertisement in shell scripts or as a diagnostic tool.

.SH OPTIONS
.TP
.BI \-v|\-\-version 
displays version of slptool and OpenSLP.
.TP
.BI \-s|\-\-scope " scopes"
.I scopes
is a comma separated list of scopes applicable for the command.
.TP
.BI [\-l|\-\-language " languagetag"
.I languagetag
is the language tag to be used for the command.
.TP
.BI findsrvs " service\-type [filter]"
.TP
.BI findattrs " url [attrids]"
.TP
.BI findsrvtypes " [authority] "
.TP
.BI findscopes
.TP
.BI register " url [attrs]"
.TP
.BI deregister " url"
.TP
.BI getproperty " propertyname"

.SH EXAMPLES
slptool register service:myserv.x://myhost.com "(attr1=val1),(attr2=val2)"
.LP
slptool findsrvs service:myserv.x
.LP
slptool findsrvs service:myserv.x "(attr1=val1)"
.LP
slptool findattrs service:myserv.x://myhost.com
.LP
slptool findattrs service:myserv.x://myhost.com attr1
.LP
slptool deregister service:myserv.x://myhost.com
.LP
slptool getproperty net.slp.useScopes

.SH SEE ALSO
"OpenSLP User's Guide"

.SH NOTES
This man page was written for the Debian project by Debian Developer Ganesan
Rajagopal (rganesan@debian.org). You should refer to the online html
documentation in the 
.BR openslp-doc 
package for more complete and up-to-date information.

.SH BUGS
http://sourceforge.net/tracker/?group_id=1730&atid=101730

.SH ACKNOWLEDGEMENTS
.B OpenSLP
is developed and maintained by The OpenSLP Project (http://www.openldap.org/).


openslp/NEWS
NEWS - list of user-visible changes between releases of OpenSLP

New in openslp-0.8.2 (2001-02-1?)
---------------------------------
* Minor fixes in the html documentation.

New in openslp-0.8.1 (2000-12-19)
---------------------------------
* slpd now correctly refreshes and ages services as required by DA and SA 
  specifications
* Location of slpd.pid and slpd.log files now depends on the configure prefix 
  on Unix.
* slpd running as a DA now sends passive DAAdverts for SLPv1 also. 
* slpd now works (again) on 64-bit platforms. 

New in openslp-0.8.0 (2000-11-30)
--------------------------------
* New OpenSLP License.  OpenSLP is now licensed under the BSD license.
* OpenSLP contrib tarball is now available.  OpenSLP Contrib is a collection
  of SLP related utilities and tools.
* Predicate support now enabled by default.
* SLPFindScopes() is now implemented to RFC spec.
* Implemented SLPFindScopes() in libslp.
* Send dead DAAdvert as part of graceful shutdown.
* DAs now echo local registrations to other DAs (slpd as a DA also acts like
  an SA for local API callers).
* Added code to ensure we respond correctly (with AUTHENICATION_UNKNOWN)
  when a request with a SPI is received.
* Changed SLPD behaviour to be less destructive on SIGHUP.  SIGHUP no longer
  dumps the entire registration database on SIGHUP.  SIGHUP only causes slpd
  to re-read configuration and static registration as well as perform DA
  discovery. 


New in openslp-0.7.8 (2000-11-06)
---------------------------------
* slpd now does a graceful shutdown on SIGTERM to deregister serviced
  with DAs
* Considerable time spend using commercial debugging tools to remove
  ALL memory leaks from slpd (without predicates) and libslp.
* IPC (via loopback) is now used by libslp to obtain DA information
  from slpd before multicasting
* slpd running as a DA now sends passive DAAdverts
* slpd running as an SA now performs active DA discovery
* slpd running as an SA now re-registers all services with new DAs
* Significant fixes were made to make slpd work properly on multi-homed machines  


New in openslp-0.7.7: (2000-10-11)
----------------------------------
* OpenSLP now supports SLPv1 when run as a DA.
* TCP stream reusage was added to libslp and slpd
* Correction of many active DA discovery problems were made in slpd
* Predicate features from 0.7.6 are not build by default because of a
  few problems that will be corrected for 0.7.8         


New in openslp-0.7.6: (2000-9-28)
----------------------------------
* OpenSLP ported to FreeBSD. Compaq Tru64 support is also complete.
* Full predicate support has been added to OpenSLP.  slpd will now correctly 
  answer service requests containing LDAP filters
* As part of the Windows port, the MSVC project files added.  This should
  make it much easier for OpenSLP to be build on Win32 platform.
* Most of slpd and libslp code was tested against a malloc debugger.  Several
  bugs and memory leaks were fixed.


New in openslp-0.7.5: (2000-09-08)
----------------------------------
* OpenSLP ported to Win32.
* SLPFindSrvTypes() now works.
* libslp minor version updated.


New in openslp-0.7.4: (2000-09-01)
----------------------------------
* Finally added SA <--> DA communication resulting in an slpd that now acts as
  a much better SA and DA.  
* slpd (as a DA) responds to active DA discovery requests.
* slpd (as an SA) performs active DA discovery, supports passive DA discovery,
  supports static DA configuration, registers and de-registers services with 
  DAs.
* libslp now has the ability to register services with to the local slpd
  (SA), or a discovered DA.  This eliminates the need to have slpd running on
  all machines in certain cases. libslp can now request services from a DA 
  or via multicast or broadcast.
  

New in openslp-0.7.3:
---------------------
* openslp-0.7.3 will not be released.


New in openslp-0.7.2: (2000-08-17)
----------------------------------
* Revamped build system.
  Takes advantage of libtool/automake integration.
  libslp versioning changed to be compatible with libtool.
  Building with non-GCC compilers is now possible.
  Build directory can be different from source directory.
  Test suite integrated with automake, run make check to run the test suite.
* Fixed several sneaky memory leaks and memory access bugs that dealt with
  langtags in libslp 
* Modified slpd command line parameter parsing to print version and help
  information  
* Ported to Solaris (2.6/7) and Linux Alpha.
* Documentation cleaned up (code set changed to iso8859-1 and spell checked).

New in openslp-0.7.1:
---------------------
* Several bug fixes.
* Now uses Autoconf/Automake.


New in openslp-0.7.0: (2000-07-06)
----------------------------------
* Addition of NEWS file.
* Move of CHANGELOG to ChangeLog.


openslp/slptool/slptool.c
/***************************************************************************/
/*                                                                         */
/* Project:     OpenSLP command line UA wrapper                            */
/*                                                                         */
/* File:        slptool.c                                                  */
/*                                                                         */
/* Abstract:    Command line wrapper for OpenSLP                           */
/*                                                                         */
/* Requires:    OpenSLP installation                                       */
/*                                                                         */
/* Author(s):   Matt Peterson <mpeterson@caldera.com>                      */
/*                                                                         */
/* Copyright (c) 1995, 1999  Caldera Systems, Inc.                         */
/*                                                                         */
/* This program is free software; you can redistribute it and/or modify it */
/* under the terms of the GNU Lesser General Public License as published   */
/* by the Free Software Foundation; either version 2.1 of the License, or  */
/* (at your option) any later version.                                     */
/*                                                                         */
/*     This program is distributed in the hope that it will be useful,     */
/*     but WITHOUT ANY WARRANTY; without even the implied warranty of      */
/*     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the       */
/*     GNU Lesser General Public License for more details.                 */
/*                                                                         */
/*     You should have received a copy of the GNU Lesser General Public    */
/*     License along with this program; see the file COPYING.  If not,     */
/*     please obtain a copy from http://www.gnu.org/copyleft/lesser.html   */
/*                                                                         */
/*-------------------------------------------------------------------------*/
/*                                                                         */
/*     Please submit patches to maintainer of http://www.openslp.org       */
/*                                                                         */
/***************************************************************************/

#include "slptool.h"

/*=========================================================================*/
SLPBoolean mySrvTypeCallback( SLPHandle hslp, 
                              const char* srvtypes, 
                              SLPError errcode, 
                              void* cookie ) 
/*=========================================================================*/
{
    char* cpy;
    char* slider1;
    char* slider2;

    if(errcode == SLP_OK && *srvtypes)
    {
        cpy = strdup(srvtypes);
        if(cpy)
        {
            slider1 = slider2 = cpy;
            slider1 = strchr(slider2,',');
	    while(slider1)
            {
                *slider1 = 0;
                printf("%s\n",slider2);
                slider1 ++;
                slider2 = slider1;
	        slider1 = strchr(slider2,',');
            }

            /* print the final itam */
            printf("%s\n",slider2);

            free(cpy);
        }
        
    }

    return SLP_TRUE;
}


/*=========================================================================*/
void FindSrvTypes(SLPToolCommandLine* cmdline)
/*=========================================================================*/
{
    SLPError    result;
    SLPHandle   hslp;

    if(SLPOpen(cmdline->lang,SLP_FALSE,&hslp) == SLP_OK)
    {
        if(cmdline->cmdparam1)
        {
            result = SLPFindSrvTypes(hslp,
				     cmdline->cmdparam1,
				     cmdline->scopes,
				     mySrvTypeCallback,
				     0);
	}
        else
        {
	    result = SLPFindSrvTypes(hslp,
				     "*",
				     cmdline->scopes,
				     mySrvTypeCallback,
				     0);
	}
       
        if(result != SLP_OK)
        {
            printf("errorcode: %i\n",result);
        }
       
        SLPClose(hslp);
    }
}

/*=========================================================================*/
SLPBoolean myAttrCallback(SLPHandle hslp, 
                        const char* attrlist, 
                        SLPError errcode, 
                        void* cookie )
/*=========================================================================*/
{
    if(errcode == SLP_OK)
    {
        printf("%s\n",attrlist);
    }
    
    return SLP_TRUE;
}


/*=========================================================================*/
void FindAttrs(SLPToolCommandLine* cmdline)
/*=========================================================================*/
{
    SLPError    result;
    SLPHandle   hslp;

    if(SLPOpen(cmdline->lang,SLP_FALSE,&hslp) == SLP_OK)
    {

        result = SLPFindAttrs(hslp,
                             cmdline->cmdparam1,
                             cmdline->scopes,
                             cmdline->cmdparam2,
                             myAttrCallback,
                             0);
        if(result != SLP_OK)
        {
            printf("errorcode: %i\n",result);
        }
        SLPClose(hslp);
    }               
}

 
/*=========================================================================*/
SLPBoolean mySrvUrlCallback( SLPHandle hslp, 
                             const char* srvurl, 
                             unsigned short lifetime, 
                             SLPError errcode, 
                             void* cookie ) 
/*=========================================================================*/
{
    if(errcode == SLP_OK)
    {
        printf("%s,%i\n",srvurl,lifetime);
    }
    
    return SLP_TRUE;
}


/*=========================================================================*/
void FindSrvs(SLPToolCommandLine* cmdline)
/*=========================================================================*/
{
    SLPError    result;
    SLPHandle   hslp;

    if(SLPOpen(cmdline->lang,SLP_FALSE,&hslp) == SLP_OK)
    {
        result = SLPFindSrvs(hslp,
                             cmdline->cmdparam1,
                             cmdline->scopes,
                             cmdline->cmdparam2,
                             mySrvUrlCallback,
                             0);
        if(result != SLP_OK)
        {
            printf("errorcode: %i\n",result);
        }
        SLPClose(hslp);
    }               
}

/*=========================================================================*/
void FindScopes(SLPToolCommandLine* cmdline)
/*=========================================================================*/
{
    SLPError    result;
    SLPHandle   hslp;
    char*       scopes;

    if(SLPOpen(cmdline->lang,SLP_FALSE,&hslp) == SLP_OK)
    {
        result = SLPFindScopes(hslp,&scopes);
        if(result == SLP_OK)
	{
	   printf("%s\n",scopes);
	   SLPFree(scopes);
	}
       
        SLPClose(hslp);
    }               
}

void mySLPRegReport(SLPHandle hslp, SLPError errcode, void* cookie)
{
    if (errcode)
    printf("(de)registration errorcode %d\n", errcode);
}

/*=========================================================================*/
void Register(SLPToolCommandLine* cmdline)
/*=========================================================================*/
{
    SLPError    result;
    SLPHandle   hslp;
    char srvtype[80] = "", *s;
    int len = 0;

    if (strncasecmp(cmdline->cmdparam1, "service:", 8) == 0)
	len = 8;

    s = strchr(cmdline->cmdparam1 + len, ':');
    if (!s)
    {
	printf("Invalid URL: %s\n", cmdline->cmdparam1);
	return;
    }
    len = s - cmdline->cmdparam1;
    strncpy(srvtype, cmdline->cmdparam1, len);
    srvtype[len] = 0;

    if(SLPOpen(cmdline->lang,SLP_FALSE,&hslp) == SLP_OK)
    {

        result = SLPReg(hslp,
			cmdline->cmdparam1,
			SLP_LIFETIME_MAXIMUM,
			srvtype,
			cmdline->cmdparam2,
			SLP_TRUE,
			mySLPRegReport,
			0);
        if(result != SLP_OK)
        {
            printf("errorcode: %i\n",result);
        }
        SLPClose(hslp);
    }               
}

/*=========================================================================*/
void Deregister(SLPToolCommandLine* cmdline)
/*=========================================================================*/
{
    SLPError    result;
    SLPHandle   hslp;

    if(SLPOpen(cmdline->lang,SLP_FALSE,&hslp) == SLP_OK)
    {

        result = SLPDereg(hslp,
			  cmdline->cmdparam1,
			  mySLPRegReport,
			  0);
        if(result != SLP_OK)
        {
            printf("errorcode: %i\n",result);
        }
        SLPClose(hslp);
    }               
}

/*=========================================================================*/
void PrintVersion(SLPToolCommandLine* cmdline)
/*=========================================================================*/
{
    printf("slptool version = %s\n",VERSION);
    printf("OpenSLP version = %s\n", 
	   SLPGetProperty("net.slp.OpenSLPVersion"));
}


/*=========================================================================*/
int ParseCommandLine(int argc,char* argv[], SLPToolCommandLine* cmdline)
/* Returns  Zero on success.  Non-zero on error.                           */
/*=========================================================================*/
{
    int i;

    if(argc < 2)
    {
        /* not enough arguments */
        return 1;
    }

    for (i=1;i<argc;i++)
    {
        if( strcasecmp(argv[i],"-v") == 0 ||
            strcasecmp(argv[i],"--version") == 0 )
        {
            if(i < argc)
            {
                cmdline->cmd = PRINT_VERSION;
		return 0;
            }
            else
            {
                return 1;
            }
        }
        else if( strcasecmp(argv[i],"-s") == 0 ||
		 strcasecmp(argv[i],"--scopes") == 0 )
        {
            i++;
            if(i < argc)
            {
                cmdline->scopes = argv[i];
            }
            else
            {
                return 1;
            }
        }
        else if( strcasecmp(argv[i],"-l") == 0 ||
                 strcasecmp(argv[i],"--lang") == 0 )
        {
            i++;
            if(i < argc)
            {
                cmdline->lang = argv[i];
            }
            else
            {
                return 1;
            }
        }
        else if(strcasecmp(argv[i],"findsrvs") == 0)
        {
            cmdline->cmd = FINDSRVS;
            
            /* service type */
            i++;
            if(i < argc)
            {
                cmdline->cmdparam1 = argv[i];
            }
            else
            {
                return 1;
            }
            
            /* (optional) filter */
            i++;
            if(i < argc)
            {
                cmdline->cmdparam2 = argv[i];
            }
            
            break;
        }
        else if(strcasecmp(argv[i],"findattrs") == 0)
        {
            cmdline->cmd = FINDATTRS;     
                  
            /* url or service type */
            i++;
            if(i < argc)
            {
                cmdline->cmdparam1 = argv[i];
            }
            else
            {
                return 1;
            }
            
            /* (optional) attrids */
            i++;
            if(i < argc)
            {
                cmdline->cmdparam2 = argv[i];
            }
        }
        else if(strcasecmp(argv[i],"findsrvtypes") == 0)
        {
            cmdline->cmd = FINDSRVTYPES;

            /* (optional) naming authority */
            i++;
            if(i < argc)
            {
                cmdline->cmdparam1 = argv[i];
            }                             
        }
        else if(strcasecmp(argv[i],"findscopes") == 0)
        {
	    cmdline->cmd = FINDSCOPES;
	}
        else if(strcasecmp(argv[i],"register") == 0)
        {
            cmdline->cmd = REGISTER;
            
            /* url */
            i++;
            if(i < argc)
            {
                cmdline->cmdparam1 = argv[i];
            }
            else
            {
                return 1;
            }
            
            /* attrids */
            i++;
            if(i < argc)
            {
                cmdline->cmdparam2 = argv[i];
            }
	    else
	    {
		return 1;
	    }
            
            break;
        }
        else if(strcasecmp(argv[i],"deregister") == 0)
        {
            cmdline->cmd = DEREGISTER;

            /* url */
            i++;
            if(i < argc)
            {
                cmdline->cmdparam1 = argv[i];
            }
	    else
	    {
		return 1;
	    }
        }
        else
        {
            return 1;
        }    
    }

    return 0;
}


/*=========================================================================*/
void DisplayUsage()
/*=========================================================================*/
{
    printf("Usage: slptool [options] command-and-arguments \n");
    printf("   options may be:\n");
    printf("      -v (or --version) displays version of slptool and OpenSLP\n");
    printf("      -s (or --scope) followed by a comma separated list of scopes\n");
    printf("      -l (or --language) followed by a language tag\n");
    printf("   command-and-arguments may be:\n");
    printf("      findsrvs service-type [filter]\n");
    printf("      findattrs url [attrids]\n");
    printf("      findsrvtypes [authority]\n");
    printf("      findscopes\n");
    printf("      register url attrs\n");
    printf("      deregister url attrs\n");
    printf("      getproperty propertyname\n");
}


/*=========================================================================*/
int main(int argc, char* argv[])
/*=========================================================================*/
{
    int                 result;
    SLPToolCommandLine  cmdline;

    /* zero out the cmdline */
    memset(&cmdline,0,sizeof(cmdline));
    
    /* Parse the command line */
    if(ParseCommandLine(argc,argv,&cmdline) == 0)
    {
        switch(cmdline.cmd)
        {
        case FINDSRVS:
            FindSrvs(&cmdline);
            break;
        
        case FINDATTRS:
            FindAttrs(&cmdline);
            break;
        
        case FINDSRVTYPES:
            FindSrvTypes(&cmdline);
            break;
	
	case FINDSCOPES:
	    FindScopes(&cmdline);
	    break;
        
        case GETPROPERTY:
//            GetProperty(&cmdline);
            break;
        case REGISTER:
            Register(&cmdline);
            break;
        case DEREGISTER:
            Deregister(&cmdline);
            break;
	 case PRINT_VERSION:
	    PrintVersion(&cmdline);
        }
    }
    else
    {
        DisplayUsage();
        result = 1;
    }

    return 0;
}
